src/rust/engine/src/externs/mod.rs

```diff
@@ -332,6 +332,54 @@ pub fn call_method(value: &PyObject, method: &str, args: &[Value]) -> Result<PyO
   value.call_method(gil.python(), method, args_tuple, None)
 }
 
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum InvocationResult {
+  SuccessfulReturn(Value),
+  Exception(Failure),
+}
+
+impl From<Result<PyObject, PyErr>> for InvocationResult {
+  fn from(other: Result<PyObject, PyErr>) -> Self {
+    match other {
+      Ok(obj) => InvocationResult::SuccessfulReturn(Value::from(obj)),
+      Err(err) => InvocationResult::Exception(Failure::from_py_err(err)),
+    }
+  }
+}
+
+impl InvocationResult {
+  pub fn unsafe_extract(self) -> Result<Value, Failure> {
```

Comment:
Why is this method called `unsafe_extract`? It doesn't seem like it's doing anything unsafe

Reply:
Correct! I have converted it into a `.try_into()` impl.